### PR TITLE
feat: filter restricted runs on APIs

### PR DESCRIPTION
**Pull request by zawan-ila**
Internal tickets: [PROD-4005](https://2u-internal.atlassian.net/browse/PROD-4005) + [PROD-4006](https://2u-internal.atlassian.net/browse/PROD-4006)

Hide restricted runs across APIs by default. Also do not include them when calculating fields like course_run_statuses. If a query-param `include_restricted=<some_value>` is present, do not hide them

# Testing
1. Create restricted runs in your system and link them with courses, programs, etc.
2. Run `update_index`
3. Verify that runs are hidden on all APIs by default and are revealed on adding the query params.

### CODE REVIEW

File: course_discovery/apps/api/serializers.py [hunk 1]
```diff
@@ -12,7 +12,7 @@
 import pytz
 import waffle  # lint-amnesty, pylint: disable=invalid-django-waffle-import
 from django.contrib.auth import get_user_model
-from django.db.models import Count
+from django.db.models import Count, prefetch_related_objects
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/serializers.py]: nit: this does not seem to be in use in this file.

File: course_discovery/apps/api/utils.py [hunk 2]
```diff
@@ -198,6 +199,9 @@ def increment_character(character):
     """
     return chr(ord(character) + 1) if character != 'z' else 'a'
 
+def get_forbidden_runs(request):
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/utils.py]: We should use get_restricted_runs (and restricted_runs as a variable) for consistency. Forbidden_runs can mislead when someone else is reading the code (with little to no context). (and to add to following comment, this method is only returning restriction type(s), not actual course runs. That is also misleading.

Comment (DawoudSheraz) [inline course_discovery/apps/api/utils.py]: - nit: missing docstring.
- the method is not really returning runs but rather restriction types.

Comment (zawan-ila) [reply course_discovery/apps/api/utils.py]: I have updated some names. Hopefully it's good enough now

Comment (zawan-ila) [reply course_discovery/apps/api/utils.py]: I have updated some names. Hopefully it's good enough now

File: course_discovery/apps/api/v1/tests/test_views/test_programs.py [hunk 3]
```diff
@@ -216,6 +219,23 @@ def test_list(self):
 
         self.assert_list_results(self.list_path, expected, 26)
 
+    @pytest.mark.parametrize( "include_restriction_param", [True, False])
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/v1/tests/test_views/test_programs.py]: nit: unnecessary space after (

File: course_discovery/apps/api/v1/tests/test_views/test_programs.py [hunk 4]
```diff
@@ -216,6 +219,23 @@ def test_list(self):
...
+    def test_list_restricted_runs(self, include_restriction_param):
+        """ Verify the endpoint returns a list of all programs. """
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/v1/tests/test_views/test_programs.py]: no need for the docstring here

File: course_discovery/apps/course_metadata/search_indexes/documents/person.py [hunk 5]
```diff
@@ -47,7 +47,7 @@ def prepare_position(self, obj):
             return []
         return [position.title, position.organization_override]
 
-    def get_queryset(self):
+    def get_queryset(self, excluded_restriction_types=[]):
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/documents/person.py]: why do we need this default but unused argument?

File: course_discovery/apps/course_metadata/search_indexes/serializers/course.py [hunk 6]
```diff
@@ -12,6 +12,7 @@
 from course_discovery.apps.api import serializers as cd_serializers
 from course_discovery.apps.api.serializers import ContentTypeSerializer, CourseWithProgramsSerializer
 from course_discovery.apps.course_metadata.utils import get_course_run_estimated_hours, get_product_skill_names
+from course_discovery.apps.course_metadata.models import CourseRun, Course
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/search_indexes/serializers/course.py]: nit: unnecessary import?

**Review by DawoudSheraz — APPROVED** [review]
Played around a bit with this locally, looks good. 
One thing: search/<> endpoints are missing restriction_type field on course_run serializers.

File: course_discovery/apps/api/v1/tests/test_views/test_search.py [hunk 7]
```diff
@@ -295,14 +327,17 @@ def setUp(self):
...
         qs = ''
 
         if query:
             qs = urllib.parse.urlencode(query, True)
 
-        url = f'{path}?{qs}'
+        if path_has_params:
+            url = f'{path}&{qs}'
+        else:
+            url = f'{path}?{qs}'
```
Comment (AfaqShuaib09) [inline course_discovery/apps/api/v1/tests/test_views/test_search.py]:
```suggestion
url = f'{path}&{qs}' if path_has_params else url = f'{path}?{qs}'
```

File: course_discovery/apps/api/v1/views/search.py [hunk 8]
```diff
@@ -16,9 +16,10 @@
 from rest_framework.views import APIView
 
 from course_discovery.apps.api import serializers
-from course_discovery.apps.api.utils import update_query_params_with_body_data
+from course_discovery.apps.api.utils import update_query_params_with_body_data, get_excluded_restriction_types
 from course_discovery.apps.course_metadata.choices import ProgramStatus
 from course_discovery.apps.course_metadata.models import Person
+
```
Comment (AfaqShuaib09) [inline course_discovery/apps/api/v1/views/search.py]: nit: unexpected extra line here

File: course_discovery/apps/api/v1/views/search.py [hunk 9]
```diff
@@ -101,6 +102,8 @@ class CourseSearchViewSet(BaseElasticsearchDocumentViewSet):
         'subjects': {'field': 'subjects.raw', 'enabled': True},
         'prerequisites': {'field': 'prerequisites', 'enabled': True},
     }
+    
+
```
Comment (AfaqShuaib09) [inline course_discovery/apps/api/v1/views/search.py]: nit: remove extra spaces

File: course_discovery/apps/api/v1/views/search.py [hunk 10]
```diff
@@ -132,6 +135,13 @@ class CourseRunSearchViewSet(FacetQueryFieldsMixin, BaseElasticsearchDocumentVie
         },
     }
 
+    def get_queryset(self):
+        """Get queryset."""
```
Comment (AfaqShuaib09) [inline course_discovery/apps/api/v1/views/search.py]: nit: docstring can be more meaningful here

File: course_discovery/apps/course_metadata/algolia_models.py [hunk 11]
```diff
@@ -87,7 +87,8 @@ def _wrap(self, *args, **kwargs):
 
 
 def get_course_availability(course):
-    all_runs = course.course_runs.filter(status=CourseRunStatus.Published)
+    all_runs = course.course_runs.all()
+    all_runs = filter(lambda r: r.status == CourseRunStatus.Published, all_runs)
```
Comment (AfaqShuaib09) [inline course_discovery/apps/course_metadata/algolia_models.py:110]: wondering why this change has been done?

File: course_discovery/apps/course_metadata/search_indexes/documents/learner_pathway.py [hunk 12]
```diff
@@ -50,10 +53,10 @@ def prepare_partner(self, obj):
     def prepare_published(self, obj):
         return obj.status == PathwayStatus.Active
 
-    def get_queryset(self):
+    def get_queryset(self, excluded_restriction_types=[]):
```
Comment (AfaqShuaib09) [inline course_discovery/apps/course_metadata/search_indexes/documents/learner_pathway.py]: nit: seems like an unused argument

**Review by AfaqShuaib09 — APPROVED** [review]
Did some playaround with the `include_restricted` in api/v1 endpoints & search endpoints on local. Changes look good.

[thinking] One thing we can do in a followup PR to check for N+1 queries.

File: course_discovery/apps/api/v1/views/search.py (hunk 10 above)
Comment (zawan-ila) [reply course_discovery/apps/api/v1/views/search.py]: I removed it altogether. I think it is obvious what `get_queryset` does.

File: course_discovery/apps/course_metadata/algolia_models.py (hunk 11 above)
Comment (zawan-ila) [reply course_discovery/apps/course_metadata/algolia_models.py:110]: We have prefetched `course_runs` on `course`. And in that prefetch, we have [excluded restricted runs](https://github.com/openedx/course-discovery/pull/4356/files#diff-1fc91f72e5dc28a966b180a151e35ea3bcc6f85236881faa99492c05ec38067aR238). Now if we dont make this change, it will go to the database and `course.course_runs` will return all runs (even restricted ones). This makes sure that the prefetched course runs are used.

File: course_discovery/apps/course_metadata/search_indexes/documents/person.py (hunk 5 above)
Comment (zawan-ila) [reply course_discovery/apps/course_metadata/search_indexes/documents/person.py]: Due to [this](https://github.com/openedx/course-discovery/pull/4356/files#diff-d6a4e1fc1688e70aaec3007cf12d5ff67736d2d992873083cd82fc15526ba09bR54) change, the get_queryset method is called with the excluded_restriction_types for all documents. We don't really use `excluded_restriction_type` in the `get_queryset` of some documents but the argument is still passed. Hence, this is to prevent an error.

We could have a condition to check the document type, and only pass it in for those documents which need it, but I think it is neater this way

File: course_discovery/apps/course_metadata/search_indexes/documents/learner_pathway.py (hunk 12 above)
Comment (zawan-ila) [reply course_discovery/apps/course_metadata/search_indexes/documents/learner_pathway.py]: Due to [this](https://github.com/openedx/course-discovery/pull/4356/files#diff-d6a4e1fc1688e70aaec3007cf12d5ff67736d2d992873083cd82fc15526ba09bR54) change, the get_queryset method is called with the excluded_restriction_types for all documents. We don't really use `excluded_restriction_type` in the `get_queryset` of some documents but the argument is still passed. Hence, this is to prevent an error.

We could have a condition to check the document type, and only pass it in for those documents which need it, but I think it is neater this way

File: course_discovery/apps/course_metadata/search_indexes/documents/person.py (hunk 5 above)
Comment (DawoudSheraz) [reply course_discovery/apps/course_metadata/search_indexes/documents/person.py]: Right. It would be good to have this context in docstrings/context for future's sake.

File: course_discovery/apps/api/v1/tests/test_views/test_catalogs.py [hunk 13]
```diff
@@ -335,6 +335,32 @@ def test_courses(self, state):
...
+        RestrictedCourseRunFactory(course_run=restricted_course_run, restriction_type='custom-b2b-enterprise')
+        SeatFactory.create(course_run=course_run)
+        SeatFactory.create(course_run=restricted_course_run)
+
+        if include_restriction_param:
+            url += '?include_restricted=custom-b2b-enterprise'
+
+        response = self.client.get(url)
+        assert response.status_code == 200
+        assert len(response.data['results'][0]['course_runs']) == 2 if include_restriction_param else 1
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/v1/tests/test_views/test_catalogs.py]: nit: this 2 and 1 (expected_result_count) can be moved to DDT alongside include_restriction_param. It will fit in better as the test is already using ddt.

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 14]
```diff
@@ -1211,6 +1211,40 @@ def test_list_sorted_by_course_start_date(self):
...
+        RestrictedCourseRunFactory(course_run=restricted_run, restriction_type='custom-b2c')
+        url = reverse('api:v1:course_run-list')
+        if include_restriction_param:
+            url += '?include_restricted=custom-b2c'
+
+        with self.assertNumQueries(14, threshold=3):
+            response = self.client.get(url)
+
+        assert response.status_code == 200
+        restrieved_keys = [r['key'] for r in response.data['results']]
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: nit: retrieved

File: course_discovery/apps/api/v1/tests/test_views/test_course_runs.py [hunk 15]
```diff
@@ -1211,6 +1211,40 @@ def test_list_sorted_by_course_start_date(self):
...
+        RestrictedCourseRunFactory(course_run=restricted_run, restriction_type='custom-b2c')
+        query = 'title:Some cool title'
+        url = '{root}?q={query}'.format(root=reverse('api:v1:course_run-list'), query=query)
+        if include_restriction_param:
+            url += '?include_restricted=custom-b2c,custom-b2b-enterprise'
+
+        with self.assertNumQueries(30, threshold=3):
+            response = self.client.get(url)
+
+        assert len(response.data['results']) == 4 if include_restriction_param else 3
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/v1/tests/test_views/test_course_runs.py]: ddt comment as above also applicable here.

File: course_discovery/apps/api/v1/tests/test_views/test_courses.py [hunk 16]
```diff
@@ -278,6 +278,68 @@ def test_course_runs_are_ordered(self):
...
+        assert response.status_code == 200
+
+        if not include_restriction_param:
+            self.assertListEqual(response.data['course_run_keys'], [run_not_restricted.key])
+            self.assertListEqual(response.data['course_run_statuses'], [run_not_restricted.status])
+            self.assertEqual(len(response.data['course_runs']), 1)
+            self.assertEqual(response.data['advertised_course_run_uuid'], None)
+        else:
+            self.assertEqual(set(response.data['course_run_keys']), {run_not_restricted.key, run_restricted.key})
+            self.assertEqual(
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/v1/tests/test_views/test_courses.py:311]: if block is doing assertListEqual, this is doing assertEqual. Since the context is same, we should try to keep the code consistent where applicable.

File: course_discovery/apps/api/v1/tests/test_views/test_programs.py [hunk 17]
```diff
@@ -48,13 +48,16 @@ def setup(self, client, django_assert_num_queries, partner):
...
         organizations = [OrganizationFactory(partner=self.partner)]
         person = PersonFactory()
 
         if courses is None:
             courses = [CourseFactory(partner=self.partner)]
-            CourseRunFactory(course=courses[0], staff=[person])
+            course_run = CourseRunFactory(course=courses[0], staff=[person])
+
+        if include_restricted_run:
+            RestrictedCourseRunFactory(course_run=course_run, restriction_type='custom-b2c')  # pylint: disable=possibly-used-before-assignment
```
Comment (DawoudSheraz) [inline course_discovery/apps/api/v1/tests/test_views/test_programs.py]: nit: if we move this `if include_restricted_run:` under if block at 55, will we still need pylint disable checks there? Considering we can't restriction without run, it makes sense to nest this if block.

File: course_discovery/apps/course_metadata/tests/test_algolia_models.py [hunk 18]
```diff
@@ -213,6 +213,17 @@ def test_should_index(self):
         course.authoring_organizations.add(OrganizationFactory())
         assert course.should_index
 
+    def test_restricted_run_should_not_index(self):
+        '''
+        Test that if a course has only one run and that run is restricted,
+        it will not be indexed
+        '''
```
Comment (DawoudSheraz) [inline course_discovery/apps/course_metadata/tests/test_algolia_models.py]: nit: edX developer docstring guideline is to use `"` not `'`.

File: course_discovery/apps/api/v1/tests/test_views/test_programs.py (hunk 17 above)
Comment (zawan-ila) [reply course_discovery/apps/api/v1/tests/test_views/test_programs.py]: I have nested the block, and have also removed the pylint disable.

File: course_discovery/apps/api/v1/tests/test_views/test_courses.py (hunk 16 above)
Comment (zawan-ila) [reply course_discovery/apps/api/v1/tests/test_views/test_courses.py:311]: I have converted both to assertEqual.

**Review by DawoudSheraz — APPROVED** [review]

**Review by AfaqShuaib09 — APPROVED** [review]
